### PR TITLE
[Gekidou] Saves groups + group-channel for constrained channels

### PR DESCRIPTION
**Pull request by shazm**
#### Summary

If the channel is constrained (i.e. for groups via LDAP etc), it fetches the groups associated to the channel and saves them accordingly on channel switch.

Note - this is waiting for https://github.com/mattermost/mattermost-mobile/pull/6357 to be merged in first.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-44671

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

**Comment by avinashlng1080** [comment]
@shazm  - see Elias comment above - perhaps you can move the below into the remote action calling this `switchToChannel` function call.


```
    if (channel.isGroupConstrained) {
                    fetchGroupsForChannel(serverUrl, channelId);
                }
```

File: app/database/operator/server_data_operator/handlers/group.ts [hunk 1]
```diff
@@ -185,6 +187,75 @@ const GroupHandler = (superclass: any) => class extends superclass implements Gr
...
+        if (groups?.length && !existingGroupChannels.length) {
+            rawValues = groups.map((g) => ({id: `${g.id}-${channelId}`, channel_id: channelId, group_id: g.id}));
+        }
+
+        // If both, we only want to save new ones and delete one's no longer in groups
+        if (groups?.length && existingGroupChannels.length) {
+            const groupsSet: {[key: string]: GroupChannel} = {};
+
+            for (const g of groups) {
+                groupsSet[`${g.id}`] = {id: `${g.id}-${channelId}`, channel_id: channelId, group_id: g.id};
```
Comment (larkox) [inline app/database/operator/server_data_operator/handlers/group.ts:227]: What is the difference between ` ${g.id}` and just using `g.id`?

Comment (shazm) [reply app/database/operator/server_data_operator/handlers/group.ts:227]: Ah typescript used to complain a few versions ago about indexing by string on Array's, old habit - have updated.

File: app/database/operator/server_data_operator/transformers/group.ts [hunk 2]
```diff
@@ -98,3 +100,29 @@ export const transformGroupTeamRecord = ({action, database, value}: TransformerA
...
+   * @param {Database} operator.database
+   * @param {RecordPair} operator.value
+   * @returns {Promise<GroupChannelModel>}
+   */
+export const transformGroupChannelRecord = ({action, database, value}: TransformerArgs): Promise<GroupChannelModel> => {
+    const raw = value.raw as GroupChannel;
+
+    // id of group comes from server response
+    const fieldsMapper = (model: GroupChannelModel) => {
+        model._raw.id = raw.id || `${raw.group_id}-${raw.channel_id}`;
```
[the remark this answers is not in the document]
Comment (shazm) [reply app/database/operator/server_data_operator/transformers/group.ts:116]: Updated.

File: app/actions/remote/channel.ts [hunk 3]
```diff
@@ -1107,6 +1108,12 @@ export async function switchToChannelById(serverUrl: string, channelId: string,
     markChannelAsRead(serverUrl, channelId);
     fetchChannelStats(serverUrl, channelId);
 
+    getChannelById(database, channelId).then((channel) => {
```
Comment (shazm) [inline app/actions/remote/channel.ts:1111]: @avinashlng1080 / @enahum - I've moved the `fetchGroupsForChannel` to the remote channel switch action, but I only need to call it if the channel is group constrained - it made more sense having it in the local action as the `ChannelModel` was already available there but do check this and let me know if it works or if there's a better way -

**Review by larkox — APPROVED** [review]

File: app/actions/remote/channel.ts [hunk 4]
```diff
@@ -1107,6 +1108,12 @@ export async function switchToChannelById(serverUrl: string, channelId: string,
...
+        if (channel?.isGroupConstrained) {
+            fetchGroupsForChannel(serverUrl, channelId);
```
Comment (enahum) [inline app/actions/remote/channel.ts:1113]: The check about constraint or not can happen inside the action in itself

**Review by enahum — CHANGES_REQUESTED** [review]

**Review by avinashlng1080 — APPROVED** [review]

File: app/actions/remote/channel.ts (hunk 4 above)
Comment (shazm) [reply app/actions/remote/channel.ts]: Yup updated to handle like I do for team switch 👍

**Review by larkox — APPROVED** [review]

**Review by avinashlng1080 — APPROVED** [review]

**Comment by avinashlng1080** [comment]
@shazm please resolve conflict

**Comment by shazm** [comment]
> @shazm please resolve conflict

@avinashlng1080 resolved 👍  cc @larkox

**Review by avinashlng1080 — APPROVED** [review]

File: app/database/operator/server_data_operator/handlers/group.ts [hunk 5]
```diff
@@ -48,6 +50,64 @@ const GroupHandler = (superclass: any) => class extends superclass implements Gr
         });
     };
 
+    /**
+     * handleGroupChannelsForChannel: Handler responsible for the Create/Update operations occurring on the GroupChannel table from the 'Server' schema
+     *
+     * @param {HandleGroupChannelsForChannelArgs}
+     * @throws DataOperatorException
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/group.ts:57]:
```suggestion
```

**Review by enahum — APPROVED** [review]

File: app/database/operator/server_data_operator/handlers/group.ts (hunk 5 above)
Comment (shazm) [reply app/database/operator/server_data_operator/handlers/group.ts]: :pray:

**Pull request merged**